File: backend/src/routes/analytics.ts
import { Router, Response } from 'express';
import mongoose from 'mongoose';
import { authenticate, AuthRequest } from '../middleware/auth';
import { Session } from '../models/Session';
import { User } from '../models/User';

const router = Router();
router.use(authenticate);

router.get('/dashboard', async (req: AuthRequest, res: Response): Promise<void> => {
  const companyId = new mongoose.Types.ObjectId(req.companyId);
  const userId = new mongoose.Types.ObjectId(req.userId);
  const isAgent = req.userRole === 'AGENT';

  const matchFilter = isAgent ? { companyId, userId } : { companyId };

  const [totalSessions, sessions] = await Promise.all([
    Session.countDocuments({ ...matchFilter, status: 'COMPLETED' }),
    Session.find({ ...matchFilter, status: 'COMPLETED' })
      .sort({ endedAt: -1 })
      .limit(20)
      .populate<{ userId: InstanceType<typeof User> }>('userId', 'firstName lastName avatarUrl')
      .populate('personaId', 'name emoji')
      .lean(),
  ]);

  const scores = sessions.map((s) => s.totalScore ?? 0).filter((s) => s > 0);
  const avgScore = scores.length ? Math.round(scores.reduce((a, b) => a + b, 0) / scores.length) : 0;

  const passRate = scores.length
    ? Math.round((scores.filter((s) => s >= 70).length / scores.length) * 100)
    : 0;

  const activeUsers = isAgent
    ? 1
    : await User.countDocuments({ companyId, isActive: true });

  const recentSessions = sessions.slice(0, 5).map((s) => {
    const u = s.userId as InstanceType<typeof User>;
    const p = s.personaId as Record<string, unknown> | null;
    return {
      id: s._id.toString(),
      endedAt: s.endedAt,
      durationSeconds: s.durationSeconds,
      framework: s.framework,
      sessionType: s.type,
      personaName: (p as { name?: string })?.name ?? '',
      personaEmoji: (p as { emoji?: string })?.emoji ?? '',
      userFirstName: u?.firstName ?? '',
      userLastName: u?.lastName ?? '',
      totalScore: s.totalScore,
    };
  });

  // Framework stats aggregation
  const frameworkAgg = await Session.aggregate([
    { $match: { ...matchFilter, status: 'COMPLETED' } },
    { $unwind: '$frameworkScores' },
    {
      $group: {
        _id: '$frameworkScores.component',
        avgScore: { $avg: '$frameworkScores.score' },
        count: { $sum: 1 },
      },
    },
    { $project: { component: '$_id', avgScore: 1, count: 1, _id: 0 } },
  ]);

  let agentExtra;
  if (isAgent) {
    const weekAgo = new Date(Date.now() - 7 * 86_400_000);
    const sessionsThisWeek = await Session.countDocuments({
      userId,
      status: 'COMPLETED',
      endedAt: { $gte: weekAgo },
    });

    // Simple streak: consecutive days with sessions
    const streak = await computeStreak(userId.toString());

    const leaderboard = await Session.aggregate([
      { $match: { companyId, status: 'COMPLETED' } },
      { $group: { _id: '$userId', avgScore: { $avg: '$totalScore' } } },
      { $sort: { avgScore: -1 } },
    ]);
    const rank = leaderboard.findIndex((e) => e._id.toString() === userId.toString()) + 1;

    agentExtra = { sessionsThisWeek, streak, rank: rank || undefined };
  }

  res.json({
    totalSessions,
    avgScore,
    activeUsers,
    passRate,
    recentSessions,
    frameworkStats: frameworkAgg,
    agentExtra,
  });
});

router.get('/leaderboard', async (req: AuthRequest, res: Response): Promise<void> => {
  const companyId = new mongoose.Types.ObjectId(req.companyId);

  const agg = await Session.aggregate([
    { $match: { companyId, status: 'COMPLETED' } },
    {
      $group: {
        _id: '$userId',
        avgScore: { $avg: '$totalScore' },
        sessionCount: { $sum: 1 },
      },
    },
    { $sort: { avgScore: -1 } },
    { $limit: 20 },
  ]);

  const userIds = agg.map((e) => e._id);
  const users = await User.find({ _id: { $in: userIds } }).lean();
  const userMap = new Map(users.map((u) => [u._id.toString(), u]));

  const result = agg.map((e, i) => {
    const u = userMap.get(e._id.toString());
    return {
      rank: i + 1,
      user: u
        ? { id: u._id.toString(), firstName: u.firstName, lastName: u.lastName, avatarUrl: u.avatarUrl }
        : { id: e._id.toString(), firstName: 'Unknown', lastName: '', avatarUrl: undefined },
      avgScore: Math.round(e.avgScore ?? 0),
      sessionCount: e.sessionCount,
    };
  });

  res.json(result);
});

async function computeStreak(userId: string): Promise<number> {
  const sessions = await Session.find({ userId, status: 'COMPLETED' })
    .sort({ endedAt: -1 })
    .limit(30)
    .lean();

  if (!sessions.length) return 0;

  const days = new Set(sessions.map((s) => {
    const d = new Date(s.endedAt ?? s.createdAt);
    return `${d.getFullYear()}-${d.getMonth()}-${d.getDate()}`;
  }));

  let streak = 0;
  const today = new Date();
  for (let i = 0; i < 30; i++) {
    const d = new Date(today);
    d.setDate(today.getDate() - i);
    const key = `${d.getFullYear()}-${d.getMonth()}-${d.getDate()}`;
    if (days.has(key)) streak++;
    else if (i > 0) break;
  }
  return streak;
}

export default router;

File: roleplayombined/bolt/working/NeoRoleplaybackend-main/app/routers/analytics.py
import asyncio
import json
from datetime import datetime, timezone, timedelta

from fastapi import APIRouter, Query
from sqlalchemy import select, func
from sqlalchemy.orm import selectinload

from app.deps import DB, CurrentUser
from app.models.session import Session, FrameworkScore
from app.models.user import User
from app.models.enums import UserRole, SessionStatus
from app.schemas.analytics import (
    DashboardResponse,
    FrameworkStatItem,
    RecentSession,
    AgentDashboardExtra,
    TeamMemberSummary,
    ManagerDashboardExtra,
    LeaderboardEntry,
    LeaderboardUser,
)

router = APIRouter(prefix="/analytics", tags=["analytics"])


def _persona_display(s: Session) -> tuple[str, str]:
    """Return (displayName, displayEmoji) from persona or scenarioContext."""
    if s.persona:
        return s.persona.name, s.persona.emoji
    if s.scenarioContext:
        try:
            sc = json.loads(s.scenarioContext)
            return sc.get("displayName", "Custom Scenario"), sc.get("displayEmoji", "🎯")
        except Exception:
            pass
    return "Custom Scenario", "🎯"


@router.get("/dashboard", response_model=DashboardResponse)
async def dashboard(current_user: CurrentUser, db: DB):
    role = current_user.role
    base_filter = [Session.companyId == current_user.companyId]
    if role == UserRole.AGENT:
        base_filter.append(Session.userId == current_user.id)

    session_subq = select(Session.id).where(*base_filter).scalar_subquery()

    agg_q = db.execute(
        select(
            func.count(Session.id).label("total"),
            func.avg(Session.totalScore).label("avg_score"),
        ).where(*base_filter)
    )
    active_q = db.execute(
        select(func.count(User.id)).where(
            User.companyId == current_user.companyId,
            User.isActive == True,
        )
    )
    pass_q = db.execute(
        select(func.count(Session.id)).where(*base_filter, Session.totalScore >= 70)
    )
    fw_q = db.execute(
        select(
            FrameworkScore.component,
            func.avg(FrameworkScore.score).label("avg_score"),
            func.count(FrameworkScore.id).label("count"),
        )
        .where(FrameworkScore.sessionId.in_(session_subq))
        .group_by(FrameworkScore.component)
        .order_by(func.avg(FrameworkScore.score).desc())
    )
    recent_q = db.execute(
        select(Session)
        .options(selectinload(Session.user), selectinload(Session.persona))
        .where(*base_filter, Session.status == SessionStatus.COMPLETED)
        .order_by(Session.endedAt.desc())
        .limit(6)
    )

    agg_result, active_result, pass_result, fw_result, recent_result = await asyncio.gather(
        agg_q, active_q, pass_q, fw_q, recent_q
    )

    agg = agg_result.one()
    total_sessions = agg.total or 0
    avg_score = float(agg.avg_score) if agg.avg_score else None
    active_users = active_result.scalar_one() or 0
    passing = pass_result.scalar_one() or 0
    pass_rate = round((passing / total_sessions * 100), 1) if total_sessions else 0.0

    framework_stats = [
        FrameworkStatItem(
            component=row.component,
            avgScore=float(row.avg_score) if row.avg_score else None,
            count=row.count,
        )
        for row in fw_result
    ]

    recent_sessions = [
        RecentSession(
            id=s.id,
            endedAt=s.endedAt,
            durationSeconds=s.durationSeconds,
            framework=s.framework.value,
            sessionType=s.type.value,
            personaName=_persona_display(s)[0],
            personaEmoji=_persona_display(s)[1],
            userFirstName=s.user.firstName,
            userLastName=s.user.lastName,
            totalScore=s.totalScore,
        )
        for s in recent_result.scalars().all()
        if s.user
    ]

    # ── Agent extra ───────────────────────────────────────────────────────────
    agent_extra = None
    if role == UserRole.AGENT:
        week_ago = datetime.now(timezone.utc) - timedelta(days=7)

        week_q = db.execute(
            select(func.count(Session.id)).where(
                Session.userId == current_user.id,
                Session.createdAt >= week_ago,
                Session.status == SessionStatus.COMPLETED,
            )
        )
        streak_q = db.execute(
            select(Session.totalScore)
            .where(
                Session.userId == current_user.id,
                Session.status == SessionStatus.COMPLETED,
                Session.totalScore.isnot(None),
            )
            .order_by(Session.endedAt.desc())
            .limit(20)
        )
        rank_q = db.execute(
            select(Session.userId, func.avg(Session.totalScore).label("avg"))
            .where(
                Session.companyId == current_user.companyId,
                Session.status == SessionStatus.COMPLETED,
                Session.totalScore.isnot(None),
            )
            .group_by(Session.userId)
            .order_by(func.avg(Session.totalScore).desc())
        )

        w_r, s_r, r_r = await asyncio.gather(week_q, streak_q, rank_q)

        sessions_this_week = w_r.scalar_one() or 0

        scores = list(s_r.scalars().all())
        streak = 0
        for score in scores:
            if score >= 70:
                streak += 1
            else:
                break

        rank = None
        for i, row in enumerate(r_r.all(), 1):
            if row.userId == current_user.id:
                rank = i
                break

        agent_extra = AgentDashboardExtra(
            sessionsThisWeek=sessions_this_week,
            streak=streak,
            rank=rank,
        )

    # ── Manager / admin extra ─────────────────────────────────────────────────
    manager_extra = None
    if role in {UserRole.MANAGER, UserRole.COMPANY_ADMIN, UserRole.SUPER_ADMIN}:
        week_ago = datetime.now(timezone.utc) - timedelta(days=7)

        if role == UserRole.MANAGER:
            members_cond = (User.managerId == current_user.id, User.isActive == True)
        else:
            members_cond = (
                User.companyId == current_user.companyId,
                User.isActive == True,
                User.id != current_user.id,
            )

        members_q = db.execute(select(User).where(*members_cond).limit(50))
        stats_q = db.execute(
            select(
                Session.userId,
                func.count(Session.id).label("total"),
                func.avg(Session.totalScore).label("avg"),
                func.max(Session.endedAt).label("last_at"),
            )
            .where(Session.companyId == current_user.companyId, Session.status == SessionStatus.COMPLETED)
            .group_by(Session.userId)
        )
        week_q = db.execute(
            select(Session.userId, func.count(Session.id).label("cnt"))
            .where(
                Session.companyId == current_user.companyId,
                Session.status == SessionStatus.COMPLETED,
                Session.createdAt >= week_ago,
            )
            .group_by(Session.userId)
        )

        mem_r, stat_r, wk_r = await asyncio.gather(members_q, stats_q, week_q)
        stats_map = {row.userId: row for row in stat_r.all()}
        week_map = {row.userId: row.cnt for row in wk_r.all()}

        summaries: list[TeamMemberSummary] = []
        for m in mem_r.scalars().all():
            st = stats_map.get(m.id)
            summaries.append(TeamMemberSummary(
                id=m.id,
                firstName=m.firstName,
                lastName=m.lastName,
                avatarUrl=m.avatarUrl,
                sessionCount=st.total if st else 0,
                avgScore=round(float(st.avg), 1) if (st and st.avg) else None,
                sessionsThisWeek=week_map.get(m.id, 0),
                lastSessionAt=st.last_at if st else None,
            ))
        summaries.sort(key=lambda x: (x.avgScore or 0), reverse=True)

        manager_extra = ManagerDashboardExtra(teamSize=len(summaries), members=summaries)

    return DashboardResponse(
        totalSessions=total_sessions,
        avgScore=avg_score,
        activeUsers=active_users,
        passRate=pass_rate,
        frameworkStats=framework_stats,
        recentSessions=recent_sessions,
        agentExtra=agent_extra,
        managerExtra=manager_extra,
    )


@router.get("/leaderboard", response_model=list[LeaderboardEntry])
async def leaderboard(
    current_user: CurrentUser,
    db: DB,
    period: str = Query("all", pattern="^(all|month)$"),
):
    filters = [Session.companyId == current_user.companyId]
    if period == "month":
        month_ago = datetime.now(timezone.utc) - timedelta(days=30)
        filters.append(Session.createdAt >= month_ago)

    result = await db.execute(
        select(
            Session.userId,
            func.avg(Session.totalScore).label("avg_score"),
            func.count(Session.id).label("session_count"),
        )
        .where(*filters, Session.totalScore.isnot(None))
        .group_by(Session.userId)
        .order_by(func.avg(Session.totalScore).desc())
        .limit(20)
    )
    rows = result.all()
    if not rows:
        return []

    user_ids = [r.userId for r in rows]
    user_result = await db.execute(select(User).where(User.id.in_(user_ids)))
    users_map = {u.id: u for u in user_result.scalars().all()}

    entries = []
    for rank, row in enumerate(rows, 1):
        u = users_map.get(row.userId)
        if not u:
            continue
        entries.append(LeaderboardEntry(
            rank=rank,
            user=LeaderboardUser(id=u.id, firstName=u.firstName, lastName=u.lastName, avatarUrl=u.avatarUrl),
            avgScore=round(float(row.avg_score), 1),
            sessionCount=row.session_count,
        ))
    return entries
